marlowe/src/Language/Marlowe/SemanticsDeserialisation.hs

```diff
@@ -8,152 +9,468 @@ import           Language.Marlowe.SemanticsTypes  (Action (..), Bound (..), Case
+                else
+                  ( if y == 1
+                      then Just (Role (TokenName z), t2)
+                      else Nothing
+                  )
+            )
+      )
 
 byteStringToChoiceId :: BuiltinByteString -> Maybe (ChoiceId, BuiltinByteString)
-byteStringToChoiceId x = do (cn, t1) <- getByteString x
-                            (co, t2) <- byteStringToParty t1
-                            return (ChoiceId cn co, t2)
+byteStringToChoiceId x =
+  case getByteString x of
+    Nothing -> Nothing
+    Just (cn, t1) ->
+      ( case byteStringToParty t1 of
+          Nothing       -> Nothing
+          Just (co, t2) -> Just (ChoiceId cn co, t2)
+      )
 
 byteStringToValueId :: BuiltinByteString -> Maybe (ValueId, BuiltinByteString)
-byteStringToValueId x = do (n, t) <- getByteString x
-                           return (ValueId n, t)
+byteStringToValueId x = case getByteString x of
+  Nothing     -> Nothing
+  Just (n, t) -> Just (ValueId n, t)
 
 byteStringToToken :: BuiltinByteString -> Maybe (Token, BuiltinByteString)
-byteStringToToken x = do (cs, t1) <- getByteString x
-                         (tn, t2) <- getByteString t1
-                         return (Token (CurrencySymbol cs) (TokenName tn), t2)
+byteStringToToken x =
+  case getByteString x of
+    Nothing -> Nothing
+    Just (cs, t1) ->
+      ( case getByteString t1 of
+          Nothing       -> Nothing
+          Just (tn, t2) -> Just (Token (CurrencySymbol cs) (TokenName tn), t2)
+      )
 
 byteStringToObservation :: BuiltinByteString -> Maybe (Observation, BuiltinByteString)
-byteStringToObservation x = do (y, t1) <- byteStringToPositiveInt x
-                               case y of
-                                 0  -> withOneObs t1 NotObs
-                                 1  -> withTwoObs t1 AndObs
-                                 2  -> withTwoObs t1 OrObs
-                                 3  -> withChoiceId t1 ChoseSomething
-                                 4  -> Just (TrueObs, t1)
-                                 5  -> Just (FalseObs, t1)
-                                 6  -> withTwoValues t1 ValueGE
-                                 7  -> withTwoValues t1 ValueGT
-                                 8  -> withTwoValues t1 ValueLT
-                                 9  -> withTwoValues t1 ValueLE
-                                 10 -> withTwoValues t1 ValueEQ
-                                 _  -> Nothing
-  where withOneObs = withOne byteStringToObservation
-        withTwoObs = withTwo byteStringToObservation
-        withChoiceId = withOne byteStringToChoiceId
-        withTwoValues = withTwo byteStringToValue
-        withOne bsToOne bs f = do (z, t) <- bsToOne bs
-                                  return (f z, t)
-        withTwo bsToOne bs f = do (y, t1) <- bsToOne bs
-                                  (z, t2) <- bsToOne t1
-                                  return (f y z, t2)
+byteStringToObservation x =
```

Comment:
Another way of dealing with pattern-matching not working against integers in Plutus on-chain code is to use guards and `PlutusTx.Eq`, for example
```haskell
  | y PlutusTx.== 0 = . . .
  | y PlutusTx.== 1 = . . .
  . . .
```
The case statements here might produce smaller compiled code, however.

Reply:
Yes, I doubt that will get optimised, but I hadn't thought of that possibility. We will be replacing this when Plutus implements serialisation anyway, so it is not too bad